### ISSUE TITLE
Total CPU time is broken on M1 macs

### ISSUE THREAD

**Issue by amomchilov** [htop-dev/htop#751]
#368 fixed the CPU percentages for M1 macs, but it didn't correctly address the calculations of CPU times.

Notice the discrepancy compared to Activity Monitor:

<img width="878" alt="Screen Shot 2021-08-20 at 7 05 42 PM" src="https://user-images.githubusercontent.com/5703449/130301426-674a5bda-aac3-4453-842a-d5c7e0691554.png">

**Comment by amomchilov** [comment]
Side note:

There's a bug in libproc when running under Rosetta 2 on M1 Macs, that's present in macOS 11.5.1 (build `20G80`), but fixed in 11.5.2 (build `20G95)`

 - `mach_timebase_info` returns a factor of `1/1`, just like on Intel
 - but the values of `pti_total_user` and `pti_total_system` are the same as native M1 execution, which are native mach tick units (each of which is 3/125th’s of a nanosecond)

Thus when you take the values of `pti_total_user` and `pti_total_system`, multiply them by the factor returned by `mach_timebase_info`, you get a value in nanoseconds which is around 41.666x to small (125/3, to be precise).

In principle, you could implement a workaround:
1. add a special case to detect when running a version of macOS prior to 11.5.2, on M1 mac, under Rosetta 2
2. ignore the result of `mach_timebase_info ` completely
3. multiply by the same 125/3 factor as if `mach_timebase_info` was run natively

Though I'm not sure it's worth it, given that the fix is provided in a minor version (so nobody should be precluded from accessing it). WDYT @BenBE ?

**Comment by BenBE** [comment]
> WDYT @BenBE ?

Send a patch for this issue and we may consider it for inclusion …

**Comment by amomchilov** [comment]
I'm just gauging interest: is it worth implementing/maintaining a workaround like this for an OS bug that may-or-may-not be fixed soon?

**Comment by BenBE** [comment]
> I'm just gauging interest: is it worth implementing/maintaining a workaround like this for an OS bug that may-or-may-not be fixed soon?

There's still code around for a MacOS kernel bug when trying to read thread information … Which affected about a few versions. So, unless the workaround is a hassle to maintain, I see no real issue in implementing it.

**Comment by amomchilov** [comment]
@BenBE Here's my stab at it https://github.com/amomchilov/htop/pull/3/files

I didn't check the OS version though, that's looked a little tricky. Is there an existing utility in htop's codebase for something like this?

I see a few sysctl values that could be relevant:

```
kern.osversion: 20G95
kern.osproductversion: 11.5.2
kern.osproductversioncompat: 10.16
```

All of them would involve parsing either the hex or the sem-ver number and being able to do greater-than/less-than comparisons.

**Comment by BenBE** [comment]
> @BenBE Here's my stab at it https://github.com/amomchilov/htop/pull/3/files (I'll open a proper PR against this repo once my other 2 PRs ship)
> 

I currently see only one PR by you open. The other one authored by you shipped in 3.0.4 AFAIC. The other one should be in the 3.1.0 release according to the milestones …

> I didn't check the OS version though, that's looked a little tricky. Is there an existing utility in htop's codebase for something like this?
> 

Cf. [here](https://github.com/htop-dev/htop/blob/f839095e3b29668d080c89f3b32fb6dccff54030/darwin/DarwinProcessList.c#L53-L68) …

The callee of this function is due to the thread info query issue I mentioned.

> I see a few sysctl values that could be relevant:
> 
> ```
> kern.osversion: 20G95
> kern.osproductversion: 11.5.2
> kern.osproductversioncompat: 10.16
> ```
> 
> All of them would involve parsing either the hex or the sem-ver number and being able to do greater-than/less-than comparisons.

Looking forward to this PR. You can also draft the PR now if you like. Has the advantage that it can be already rebased as the code progresses.

**Comment by amomchilov** [comment]
> I currently see only one PR by you open. The other one authored by you shipped in 3.0.4 AFAIC. The other one should be in the 3.1.0 release according to the milestones …

Sorry Ben, I was unclear: This is the first PR i was referring to: https://github.com/htop-dev/htop/pull/752

This is (will be)  second one I was referring to, it's not yet opened: https://github.com/htop-dev/htop/pull/752#discussion_r693526482

I'll do some tidying and open these up and all that soon. I have some other stuff on my plate that I'm juggling.

Is there anything else blocking the merging of #752? Are you just waiting on a review by fasterit?

> `CompareKernelVersion`

That's perfect, thanks! I'll incorporate it later.

Should I use something like `CompareKernelVersion(999, 999, 999)` as a place holder until we know specifically which version fixes the bug?

**Comment by BenBE** [comment]
> > I currently see only one PR by you open. The other one authored by you shipped in 3.0.4 AFAIC. The other one should be in the 3.1.0 release according to the milestones …
> 
> Sorry Ben, I was unclear: This is the first PR i was referring to: #752
> 
> This is (will be) second one I was referring to, it's not yet opened: [#752 (comment)](https://github.com/htop-dev/htop/pull/752#discussion_r693526482)
> 

Ah, was just looking at the list of currently open PRs for this repo. NP. Looking forward to those two new PRs.

> I'll do some tidying and open these up and all that soon. I have some other stuff on my plate that I'm juggling.
> 

As said, maybe just open them up as draft PRs, so they already can receive some initial feedback. Depending on how the feedback round goes they might end up merged even before the current release.

> Is there anything else blocking the merging of #752? Are you just waiting on a review by fasterit?
> 

Correctly. Currently waiting for fasterit who wanted to take a look tomorrow. Due to having (and wanting) no Apple devices there's only so much I can do for review of those issues. Thus I asked fasterit for the final look at things.

> > `CompareKernelVersion`
> 
> That's perfect, thanks! I'll incorporate it later.
> 
> Should I use something like `CompareKernelVersion(999, 999, 999)` as a place holder until we know specifically which version fixes the bug?

I'd go for all of the same major+minor to have the bug, So like assuming all 11.5.x are affected, until we know more. Document this assumption (in an open ticket) and update that ticket, once we know more.

**Comment by amomchilov** [comment]
I sprouted the Rosetta 2 aspect of this into a new issue: #765
